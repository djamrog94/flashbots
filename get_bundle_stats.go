package flashbots

import (
	"math/big"
	"time"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/common/hexutil"
	"github.com/ethereum/go-ethereum/rpc"
)

type bundleStatMarshaling struct {
	BundleHash  common.Hash  `json:"bundleHash"`
	BlockNumber *hexutil.Big `json:"blockNumber"`
}

type BundleStatsResponse struct {
	IsSimulated    bool      `json:"isSimulated"`
	IsSentToMiners bool      `json:"isSentToMiners"`
	IsHighPriority bool      `json:"isHighPriority"`
	SimulatedAt    time.Time `json:"simulatedAt"`
	SubmittedAt    time.Time `json:"submittedAt"`
	SentToMinersAt time.Time `json:"sentToMinersAt"`
}

// BundleStats requests the bundles Flashbots relay stats. The given block
// number must be within 20 blocks of the current chain tip.
func BundleStats(bundleHash common.Hash, blockNumber *big.Int) *BundleStatsFactory {
	return &BundleStatsFactory{bundleHash: bundleHash, blockNumber: blockNumber}
}

type BundleStatsFactory struct {
	// args
	bundleHash  common.Hash
	blockNumber *big.Int

	// returns
	result  BundleStatsResponse
	returns *BundleStatsResponse
}

func (f *BundleStatsFactory) Returns(bundleStats *BundleStatsResponse) *BundleStatsFactory {
	f.returns = bundleStats
	return f
}

// CreateRequest implements the w3/core.RequestCreator interface.
func (f *BundleStatsFactory) CreateRequest() (rpc.BatchElem, error) {
	return rpc.BatchElem{
		Method: "flashbots_getBundleStats",
		Args: []interface{}{&bundleStatMarshaling{
			BundleHash:  f.bundleHash,
			BlockNumber: (*hexutil.Big)(f.blockNumber),
		}},
		Result: &f.result,
	}, nil
}

// HandleResponse implements the w3/core.ResponseHandler interface.
func (f *BundleStatsFactory) HandleResponse(elem rpc.BatchElem) error {
	if err := elem.Error; err != nil {
		return err
	}
	*f.returns = f.result
	return nil
}
