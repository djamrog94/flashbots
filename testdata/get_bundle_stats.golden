> {"jsonrpc":"2.0","id":1,"method":"flashbots_getBundleStats","params":[{"bundleHash":"0x2228f5d8954ce31dc1601a8ba264dbd401bf1428388ce88238932815c5d6f23f","blockNumber":"0x3b9ac9ff"}]}
< {"jsonrpc":"2.0","id":1,"result":{"isSimulated":true,"isSentToMiners":true,"isHighPriority":true,"simulatedAt":"2021-08-06T21:36:06.317Z","submittedAt":"2021-08-06T21:36:06.250Z","sentToMinersAt":"2021-08-06T21:36:06.343Z"}}
